lib/client/kube/kube.go

```diff
@@ -0,0 +1,79 @@
+// Teleport
+// Copyright (C) 2023  Gravitational, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package kube
+
+import (
+	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
+
+	"github.com/gravitational/teleport"
+	"github.com/gravitational/teleport/lib/client"
+	"github.com/gravitational/teleport/lib/tlsca"
+)
+
+var log = logrus.WithFields(logrus.Fields{
+	trace.Component: teleport.ComponentKubeClient,
+})
+
+// CheckIfCertsAreAllowedToAccessCluster evaluates if the new cert created by the user
+// to access kubeCluster has at least one kubernetes_user or kubernetes_group
+// defined. If not, it returns an error.
+// This is a safety check in order to print a better message to the user even
+// before hitting Teleport Kubernetes Proxy.
+func CheckIfCertsAreAllowedToAccessCluster(k *client.Key, rootCluster, teleportCluster, kubeCluster string) error {
+	// This is a safety check in order to print a better message to the user even
+	// before hitting Teleport Kubernetes Proxy.
+	// We only enforce this check for root clusters, since we don't have knowledge
+	// of the RBAC role mappings for remote clusters.
+	if rootCluster != teleportCluster {
+		return nil
+	}
+	for k8sCluster, cert := range k.KubeTLSCerts {
+		if k8sCluster != kubeCluster {
+			continue
+		}
+		log.Debugf("Got TLS cert for Kubernetes cluster %q", k8sCluster)
+		exist, err := checkIfCertHasKubeGroupsAndUsers(cert)
+		if err != nil {
+			return trace.Wrap(err)
+		} else if exist {
+			return nil
+		}
+	}
+	errMsg := "Your user's Teleport role does not allow Kubernetes access." +
+		" Please ask cluster administrator to ensure your role has appropriate kubernetes_groups and kubernetes_users set."
+	return trace.AccessDenied(errMsg)
```

Comment:
Since this is an access denied error, it won't get shown in the UI in Connect because lib/teleterm truncates all messages from AccessDenied errors (#32550). It'll be visible in logs though and once we address #32550, it'll make its way to the UI as well.